File: benchmarks/python/citylots.py
#!/usr/bin/env python3

import sys
import os
import logging
import time
import json

os.environ['CENTAURUS_DL_PATH'] = '../../build'
sys.path.append('../../src/python/')

from centaurus import *

logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)

class JsonListener(object):
    def parseRootObject(self, ctx):
        return ctx[0]
    def parseFeatureList(self, ctx):
        return list(ctx)
    def parseFeatureDict(self, ctx):
        if ctx:
            return json.loads(ctx.readbytes())
    def parsePropertyDict(self, ctx):
        if ctx:
            return True
    def parseTargetPropertyEntry(self, ctx):
        return True
    def defaultact(self, ctx):
        pass

if __name__ == '__main__':
    app_name = os.path.basename(sys.argv[0]).split('.')[0]
    num_workers = int(sys.argv[1]) if len(sys.argv) > 1 else 1
    logging.basicConfig(filename='debug_{}-p{}.log'.format(app_name, num_workers), filemode='w', level=logging.DEBUG)

    start_time = time.time()
    context = Context('../../grammars/citylots.cgr')
    end_time = time.time()
    gen_time = end_time - start_time
    input_path = '../../datasets/citylots.json'
    listener = JsonListener()
    context.attach(listener)
    context.start(num_workers)
    start_time = time.time()
    result = context.parse(input_path)
    end_time = time.time()
    context.stop()
    print(num_workers, end_time - start_time)

    if len(sys.argv) > 2 and sys.argv[2].lower() == 'debug':
        with open('result_{}.log'.format(app_name), 'w') as dst:
            for item in result:
                if isinstance(item, dict) and len(item) > 0:
                    print(item, file=dst)

File: benchmarks/cpp/citylots.cpp
#include "Context.hpp"
#include "rapidjson/document.h"

#include "rapidjson/ostreamwrapper.h"
#include "rapidjson/writer.h"

#include <vector>
#include <chrono>
#include <iostream>

using namespace Centaurus;

int placeholder;

void *parseRootObject(const SymbolContext<char>& ctx)
{
  return ctx.value<std::vector<rapidjson::Document*>>(1);
}

static std::vector<rapidjson::Document*> result;

void *parseFeatureList(const SymbolContext<char>& ctx)
{
  for (int i = 1; i <= ctx.count(); i++) {
    result.emplace_back(ctx.value<rapidjson::Document>(i));
  }
  return &result;
}

void *parseFeatureDict(const SymbolContext<char>& ctx)
{
  if (ctx.count() > 0) {
    auto ret = new rapidjson::Document;
    ret->Parse(ctx.read().c_str());
    return ret;
  }
  return nullptr;
}

void *parsePropertyDict(const SymbolContext<char>& ctx)
{
  if (ctx.count() > 0) {
    return &placeholder;
  }
  return nullptr;
}

void *parseTargetPropertyEntry(const SymbolContext<char>& ctx)
{
  return &placeholder;
}

int main(int argc, const char *argv[])
{
  if (argc < 2) return 1;
  int worker_num = std::atoi(argv[1]);
  bool no_action = argc >= 3 && argv[2] == std::string("dry");
  bool size = argc >= 3 && argv[2] == std::string("size");
  bool debug = argc >= 3 && argv[2] == std::string("debug");

  const char *input_path = "../../datasets/citylots.json";
  const char *grammar_path = "../../grammars/citylots.cgr";

  Context<char> context{grammar_path};

  if (!no_action) {
    context.attach(L"RootObject", parseRootObject);
    context.attach(L"FeatureList", parseFeatureList);
    context.attach(L"FeatureDict", parseFeatureDict);
    context.attach(L"PropertyDict", parsePropertyDict);
    context.attach(L"TargetPropertyEntry", parseTargetPropertyEntry);
  }

  using namespace std::chrono;

  auto start = high_resolution_clock::now();;

  context.parse(input_path, worker_num);

  auto end = high_resolution_clock::now();;

  std::cout << worker_num << " " << duration_cast<milliseconds>(end - start).count() << std::endl;

  if (size) {
    std::cout << result.size() << std::endl;
  }
  if (debug) {
    for (auto doc_ptr : result) {
      rapidjson::OStreamWrapper osw(std::cout);
      rapidjson::Writer<rapidjson::OStreamWrapper> writer(osw);
      doc_ptr->Accept(writer);
    }
    std::cout << std::endl;
  }
}
